nodebuilder/header/header.go

```diff
@@ -0,0 +1,148 @@
+package header
+
+import (
+	"context"
+	"encoding/hex"
+
+	"github.com/ipfs/go-datastore"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/peerstore"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/multiformats/go-multiaddr"
+	tmbytes "github.com/tendermint/tendermint/libs/bytes"
+
+	"github.com/celestiaorg/celestia-node/fraud"
+	"github.com/celestiaorg/celestia-node/header"
+	"github.com/celestiaorg/celestia-node/header/p2p"
+	"github.com/celestiaorg/celestia-node/header/store"
+	"github.com/celestiaorg/celestia-node/header/sync"
+	"github.com/celestiaorg/celestia-node/params"
+	headerservice "github.com/celestiaorg/celestia-node/service/header"
+)
+
+// Service creates a new header.Service.
+func Service(
+	syncer *sync.Syncer,
+	sub header.Subscriber,
+	p2pServer *p2p.ExchangeServer,
+	ex header.Exchange,
+	store header.Store,
+) *headerservice.Service {
+	return headerservice.NewHeaderService(syncer, sub, p2pServer, ex, store)
+}
+
+// P2PExchange constructs new Exchange for headers.
+func P2PExchange(cfg Config) func(params.Bootstrappers, host.Host) (header.Exchange, error) {
+	return func(bpeers params.Bootstrappers, host host.Host) (header.Exchange, error) {
+		peers, err := cfg.trustedPeers(bpeers)
+		if err != nil {
+			return nil, err
+		}
+		ids := make([]peer.ID, len(peers))
+		for index, peer := range peers {
+			ids[index] = peer.ID
+			host.Peerstore().AddAddrs(peer.ID, peer.Addrs, peerstore.PermanentAddrTTL)
+		}
+		return p2p.NewExchange(host, ids), nil
+	}
+}
+
+// InitStore initializes the store.
+func InitStore(cfg *Config) func(context.Context, params.Network, header.Store, header.Exchange) error {
+	return func(ctx context.Context, net params.Network, s header.Store, ex header.Exchange) error {
+		trustedHash, err := cfg.trustedHash(net)
+		if err != nil {
+			return err
+		}
+
+		err = store.Init(ctx, s, ex, trustedHash)
+		if err != nil {
+			// TODO(@Wondertan): Error is ignored, as otherwise unit tests for Node construction fail.
+			// 	This is due to requesting step of initialization, which fetches initial Header by trusted hash from
+			//  the network. The step can't be done during unit tests and fixing it would require either
+			//   * Having some test/dev/offline mode for Node that mocks out all the networking
+			//   * Hardcoding full extended header in params pkg, instead of hashes, so we avoid requesting step
+			log.Errorf("initializing store failed: %s", err)
+		}
+
+		return nil
+	}
+}
+
+// P2PSubscriber creates a new p2p.Subscriber.
+func P2PSubscriber(sub *pubsub.PubSub) (header.Broadcaster, header.Subscriber, *p2p.Subscriber) {
+	p2pSub := p2p.NewSubscriber(sub)
+	return p2pSub, p2pSub, p2pSub
+}
+
+// FraudService constructs fraud proof service.
+func FraudService(
```

Comment:
I'm fine to keep this here for now, but I think fraud should be its own module eventually. (as part of public api work)